R/utils.R

```diff
@@ -235,45 +255,94 @@ assert_sufficient_f_args <- function(f, ...) {
 #'
 #' @noRd
 as_slide_computation <- function(x,
+                        before,
+                        calc_ref_time_value = FALSE,
                         env = global_env(),
                         ...,
                         arg = caller_arg(x),
                         call = caller_env()) {
-  check_dots_empty0(...)
+  # A quosure is a type of formula, so be careful with the order and contents
+  # of the conditional logic here.
+  if (is_quosure(x)) {
+    if (calc_ref_time_value) {
+      f_wrapper = function(.x, .group_key, ...) {
+        .ref_time_value = min(.x$time_value) + before
+        .x <- .x[.x$.real,]
+        .x$.real <- NULL
+
+        data_env = rlang::as_environment(.x)
+        data_mask = rlang::new_data_mask(bottom = data_env, top = data_env)
+        data_mask$.data <- rlang::as_data_pronoun(data_mask)
+        # Also install `.x` directly.
+        data_mask$.x = .x
+        data_mask$.group_key = .group_key
+        data_mask$.ref_time_value = .ref_time_value
+        rlang::eval_tidy(x, data_mask)
+      }
+      return(f_wrapper)
+    }
 
-  if (is_function(x)) {
-    return(x)
+    f_wrapper = function(.x, .group_key, .ref_time_value, ...) {
+      # Convert to environment to standardize between tibble and R6
+      # based inputs. In both cases, we should get a simple
+      # environment with the empty environment as its parent.
+      data_env = rlang::as_environment(.x)
+      data_mask = rlang::new_data_mask(bottom = data_env, top = data_env)
+      data_mask$.data <- rlang::as_data_pronoun(data_mask)
+      # We'll also install `.x` directly, not as an `rlang_data_pronoun`, so
+      # that we can, e.g., use more dplyr and epiprocess operations.
+      data_mask$.x = .x
+      data_mask$.group_key = .group_key
+      data_mask$.ref_time_value = .ref_time_value
+      rlang::eval_tidy(x, data_mask)
+    }
```

Comment:
Noting again that this seems a little clunky/repetitive. These two conditional branches are identical except for the 3 lines that calcuate `ref_time_value` in the `epi_slide` case. I considered changing

```
        .ref_time_value = min(.x$time_value) + before
        .x <- .x[.x$.real,]
        .x$.real <- NULL
```

to

```
        if (calc_ref_time_value) {
          .ref_time_value = min(.x$time_value) + before
          .x <- .x[.x$.real,]
          .x$.real <- NULL
        }
```

and letting `...` capture the unused `.ref_time_value` arg in the `epi_slide` computation function.

One downside of this approach is that adding an `if` block to the `epix_slide` computation will make it somewhat (how much?) slower.

Reply:
Addressed in https://github.com/cmu-delphi/epiprocess/pull/337/commits/00e967d72ba7e50c58ac1256dc2cbc614610a257